contract/src/lib.rs

```diff
@@ -125,11 +125,37 @@ impl BtcLightClient {
         contract
     }
 
+    #[payable]
     #[pause(except(roles(Role::UnrestrictedSubmitBlocks)))]
-    pub fn submit_blocks(&mut self, #[serializer(borsh)] headers: Vec<Header>) {
+    pub fn submit_blocks(
+        &mut self,
+        #[serializer(borsh)] headers: Vec<Header>,
+    ) -> PromiseOrValue<()> {
+        let amount = env::attached_deposit();
+        let initial_storage = env::storage_usage();
+        let num_of_headers = headers.len().try_into().unwrap();
+
         for header in headers {
             self.submit_block_header(header);
         }
+
+        self.run_mainchain_gc(num_of_headers);
+
+        let diff_storage_usage = env::storage_usage().saturating_sub(initial_storage);
```

Comment:
This logic doesn't work with collections from `near_sdk::store` because it writes to the storage (flash) only after the drop is called

Reply:
It works if `flush` is called manually 
https://github.com/Near-One/btc-light-client-contract/pull/59/commits/1e5c76d84ff7316035d7214b90d84dfa31c6a80c

Reply:
Replaced store with collection here https://github.com/Near-One/btc-light-client-contract/pull/59/commits/f60c4c1efd1a5d1d4b36a4f777b6bd13b6f6d185

The collections use less gas for our scenario